MeshGeoToolsLib/GeoMapper.cpp

```diff
@@ -115,34 +138,54 @@ void GeoMapper::mapData()
 		min_val = bounding_box.getMinPoint()[2];
 		max_val = bounding_box.getMaxPoint()[2];
 	}
-	std::size_t nPoints (points->size());
 
-	if (!is_borehole)
+	std::vector<GeoLib::Point*> const* points (_geo_objects.getPointVec(_geo_name));
+	std::size_t nPoints (points->size());
+	for (unsigned j=0; j<nPoints; ++j)
 	{
-		for (unsigned j=0; j<nPoints; ++j)
+		GeoLib::Point* pnt ((*points)[j]);
+		double offset = (_grid) ? (getMeshElevation((*pnt)[0],(*pnt)[1], min_val, max_val) - (*pnt)[2])
+			                    :  getDemElevation(*pnt);
+
+		if (!GeoLib::isBorehole(pnt))
+			continue;
+		GeoLib::StationBorehole* borehole = static_cast<GeoLib::StationBorehole*>(pnt);
+		const std::vector<GeoLib::Point*> layers = borehole->getProfile();
+		std::size_t nLayers = layers.size();
+		for (unsigned k=0; k<nLayers; ++k)
 		{
-			GeoLib::Point* pnt ((*points)[j]);
-			(*pnt)[2] = (_grid) ? this->getMeshElevation((*pnt)[0],(*pnt)[1], min_val, max_val)
-				                : this->getDemElevation(*pnt);
+			GeoLib::Point* layer_pnt = layers[k];
+			(*layer_pnt)[2] = (*layer_pnt)[2] + offset;
 		}
 	}
-	else
+}
+
+void GeoMapper::mapPointDataToDEM()
+{
+	std::vector<GeoLib::Point*> const& pnts(*_geo_objects.getPointVec(_geo_name));
+
+	for (auto pnt : pnts)
```

Comment:
Caution, implicit stuff! The loop only works because `pnts` contains pointers. It would not work if `pnts` contained objects. This subtlety can be easily overlooked.
I think `for (auto* pnt : pnts)` should work here and would be better, because it stresses the fact of coping with pointers.